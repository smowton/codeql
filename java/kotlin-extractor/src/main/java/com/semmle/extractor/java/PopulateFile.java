package com.semmle.extractor.java;

import java.io.File;
import java.util.HashMap;
import java.util.Map;

import com.github.codeql.Label;
import com.github.codeql.TrapWriter;
import com.github.codeql.KotlinExtractorDbSchemeKt;
import com.semmle.util.trap.pathtransformers.PathTransformer;
import com.semmle.util.files.FileUtil;

// A temporary port of TrapWriter.populateFile et al, until we can just use the existing TrapWriter or split it up

public class PopulateFile {

  private TrapWriter tw;
  private PathTransformer transformer;
  public PopulateFile(TrapWriter tw) {
    this.tw = tw;
    this.transformer = PathTransformer.std();
  }

	private static final String[] keyReplacementMap = new String[127];
	static {
		keyReplacementMap['&'] = "&amp;";
		keyReplacementMap['{'] = "&lbrace;";
		keyReplacementMap['}'] = "&rbrace;";
		keyReplacementMap['"'] = "&quot;";
		keyReplacementMap['@'] = "&commat;";
		keyReplacementMap['#'] = "&num;";
	}

	/**
	 * Escape a string for use in a TRAP key, by replacing special characters with HTML entities.
	 * <p>
	 * The given string cannot contain any sub-keys, as the delimiters <code>{</code> and <code>}</code>
	 * are escaped.
	 * <p>
	 * To construct a key containing both sub-keys and arbitrary input data, escape the individual parts of
	 * the key rather than the key as a whole, for example:
	 * <pre>
	 * "foo;{" + label.toString() + "};" + escapeKey(data)
	 * </pre>
	 */
	public static String escapeKey(String s) {
		StringBuilder sb = null;
		int lastIndex = 0;
		for (int i = 0; i < s.length(); ++i) {
			char ch = s.charAt(i);
			switch (ch) {
			case '&':
			case '{':
			case '}':
			case '"':
			case '@':
			case '#':
				if (sb == null) {
					sb = new StringBuilder();
				}
				sb.append(s, lastIndex, i);
				sb.append(keyReplacementMap[ch]);
				lastIndex = i + 1;
				break;
			}
		}
		if (sb != null) {
			sb.append(s, lastIndex, s.length());
			return sb.toString();
		} else {
			return s;
		}
	}

  public Label populateFile(File absoluteFile) {
    String databasePath = transformer.fileAsDatabaseString(absoluteFile);
    // Ensure the rewritten path is used from now on.
    File normalisedFile = new File(databasePath);
    Label result = tw.getLabelFor("@\"" + escapeKey(databasePath) + ";sourcefile" + "\"");
    KotlinExtractorDbSchemeKt.writeFiles(tw, result, databasePath);
    populateParents(normalisedFile, result);
    return result;
  }

	private Label addFolderTuple(String databasePath) {
		Label result = tw.getLabelFor("@\"" + escapeKey(databasePath) + ";folder" + "\"");
		KotlinExtractorDbSchemeKt.writeFolders(tw, result, databasePath);
		return result;
	}

	/**
	 * Populate the parents of an already-normalised file. The path transformers
	 * and canonicalisation of {@link PathTransformer#fileAsDatabaseString(File)} will not be
	 * re-applied to this, so it should only be called after proper normalisation
	 * has happened. It will fill in all parent folders in the current TRAP file.
	 */
	private void populateParents(File normalisedFile, Label label) {
		File parent = normalisedFile.getParentFile();
		if (parent == null) return;

    Label parentLabel = addFolderTuple(FileUtil.normalisePath(parent.getPath()));
    populateParents(parent, parentLabel);
		KotlinExtractorDbSchemeKt.writeContainerparent(tw, parentLabel, label);
	}

}