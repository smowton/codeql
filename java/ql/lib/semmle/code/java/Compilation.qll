/**
 * Provides a class representing individual compiler invocations that occurred during the build.
 */

import semmle.code.FileSystem

/**
 * An invocation of the compiler. Note that more than one file may be
 * compiled per invocation. For example, this command compiles three
 * source files:
 *
 *   javac Foo.java Bar.java Baz.java
 *
 * Two things happen to each file during a compilation:
 *
 *   1. The file is compiled by a real compiler, such as javac or kotlinc.
 *   2. The file is parsed by the CodeQL front-end.
 *   3. The parsed representation is converted to database tables by
 *      the CodeQL extractor.
 *
 * This class provides CPU and elapsed time information for steps 2 and 3,
 * but not for step 1.
 */
class Compilation extends @compilation {
  /** Gets a textual representation of this element. */
  string toString() {
    exists(string name |
      compilations(this, _, _, name) and
      result = "<compilation " + name + ">"
    )
  }

  /** Gets a file compiled during this invocation. */
  File getAFileCompiled() { result = getFileCompiled(_) }

  /** Gets the `i`th file compiled during this invocation */
  File getFileCompiled(int i) { compilation_compiling_files(this, i, result) }

  /**
   * Gets the amount of CPU time spent processing file number `i` in the
   * front-end.
   */
  float getFrontendCpuSeconds(int i) { compilation_time(this, i, 1, result) }

  /**
   * Gets the amount of elapsed time while processing file number `i` in the
   * front-end.
   */
  float getFrontendElapsedSeconds(int i) { compilation_time(this, i, 2, result) }

  /**
   * Gets the amount of CPU time spent processing file number `i` in the
   * extractor.
   */
  float getExtractorCpuSeconds(int i) { compilation_time(this, i, 3, result) }

  /**
   * Gets the amount of elapsed time while processing file number `i` in the
   * extractor.
   */
  float getExtractorElapsedSeconds(int i) { compilation_time(this, i, 4, result) }

  /**
   * Gets an argument passed to the extractor on this invocation.
   */
  string getAnArgument() { result = getArgument(_) }

  /**
   * Gets the `i`th argument passed to the extractor on this invocation.
   */
  string getArgument(int i) { compilation_args(this, i, result) }

  /**
   * Gets the total amount of CPU time spent processing all the files in the
   * compiler.
   */
  float getCompilerCpuSeconds() { compilation_compiler_times(this, result, _) }

  /**
   * Gets the total amount of elapsed time while processing all the files in
   * the compiler.
   */
  float getCompilerElapsedSeconds() { compilation_compiler_times(this, _, result) }

  /**
   * Gets the total amount of CPU time spent processing all the files in the
   * front-end and extractor.
   */
  float getTotalCpuSeconds() { compilation_finished(this, result, _) }

  /**
   * Gets the total amount of elapsed time while processing all the files in
   * the front-end and extractor.
   */
  float getTotalElapsedSeconds() { compilation_finished(this, _, result) }

  /**
   * Holds if this is a compilation of Java code.
   */
  predicate isJava() { this instanceof @javacompilation }

  /**
   * Holds if this is a compilation of Kotlin code.
   */
  predicate isKotlin() { this instanceof @kotlincompilation }

  /**
   * Holds if extraction for the compilation started.
   */
  predicate extractionStarted() { compilation_started(this) }

  /**
   * Holds if the extractor terminated normally. Terminating with an exit
   * code indicating that an error occurred is considered normal
   * termination, but crashing due to something like a segfault is not.
   */
  predicate normalTermination() { compilation_finished(this, _, _) }
}
