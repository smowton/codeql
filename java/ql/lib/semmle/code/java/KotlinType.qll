/**
 * Provides classes and predicates for working with Kotlin types.
 */

import java

class KotlinType extends Element, @kt_type {
}

class KotlinNullableType extends KotlinType, @kt_nullable_type {
  override string toString() {
    exists(ClassOrInterface ci |
           kt_nullable_types(this, ci) and
           result = "Kotlin nullable " + ci.toString())
  }
  override string getAPrimaryQlClass() { result = "KotlinNullableType" }
}

class KotlinNotnullType extends KotlinType, @kt_notnull_type {
  override string toString() {
    exists(ClassOrInterface ci |
           kt_notnull_types(this, ci) and
           result = "Kotlin not-null " + ci.toString())
  }
  override string getAPrimaryQlClass() { result = "KotlinNotnullType" }
}

class KotlinTypeAlias extends Element, @kt_type_alias {
  override string getAPrimaryQlClass() { result = "KotlinTypeAlias" }

  KotlinType getKotlinType() {
    kt_type_alias(this, _, result)
  }
}
