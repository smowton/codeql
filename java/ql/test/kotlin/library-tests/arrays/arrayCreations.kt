package test

class TestArrayCreation {
  fun test1() {
    val a0 = arrayOfNulls<Int>(1)
    val a1 = arrayOf(1, 2, 3, 4)
    val a2 = doubleArrayOf(1.0, 2.0, 3.0, 4.0)
    val a3 = floatArrayOf(1.0f, 2.0f, 3.0f, 4.0f)
    val a4 = longArrayOf(1, 2, 3, 4)
    val a5 = intArrayOf(1, 2, 3, 4)
    val a6 = charArrayOf('a', 'b')
    val a7 = shortArrayOf(1, 2, 3, 4)
    val a8 = byteArrayOf(1, 2, 3, 4)
    val a9 = booleanArrayOf(true, false, true, true)

    val a10 = Array<Int>(1) { 1 }
    val a11 = Array(5) { 1 }
    val a12 = IntArray(5)
    val a13 = IntArray(5) { 1 }
    var a14 = IntArray(5) { it * 1 }
    val a15 = Array(4) { IntArray(2) }
  }
}
