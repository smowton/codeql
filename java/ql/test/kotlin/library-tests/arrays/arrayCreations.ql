import java

query predicate arrayCreationTypes(ArrayCreationExpr ace, Type t) { t = ace.getType() }

query predicate arrayCreationDimensions(ArrayCreationExpr ace, Expr dimension, int dimensionIdx) {
  ace.getDimension(dimensionIdx) = dimension
}

query predicate arrayCreationInit(ArrayCreationExpr ace, ArrayInit init, Expr e, int idx) {
  ace.getInit() = init and
  init.getInit(idx) = e
}
