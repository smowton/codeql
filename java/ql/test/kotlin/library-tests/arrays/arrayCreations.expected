arrayCreationTypes
| arrayCreations.kt:5:14:5:33 | new Integer[] | file://:0:0:0:0 | Integer[] | file://:0:0:0:0 | Integer |
| arrayCreations.kt:6:14:6:32 | new Integer[] | file://:0:0:0:0 | Integer[] | file://:0:0:0:0 | Integer |
| arrayCreations.kt:7:14:7:46 | new double[] | file://:0:0:0:0 | double[] | file://:0:0:0:0 | double |
| arrayCreations.kt:8:14:8:49 | new float[] | file://:0:0:0:0 | float[] | file://:0:0:0:0 | float |
| arrayCreations.kt:9:14:9:36 | new long[] | file://:0:0:0:0 | long[] | file://:0:0:0:0 | long |
| arrayCreations.kt:10:14:10:35 | new int[] | file://:0:0:0:0 | int[] | file://:0:0:0:0 | int |
| arrayCreations.kt:11:14:11:34 | new char[] | file://:0:0:0:0 | char[] | file://:0:0:0:0 | char |
| arrayCreations.kt:12:14:12:37 | new short[] | file://:0:0:0:0 | short[] | file://:0:0:0:0 | short |
| arrayCreations.kt:13:14:13:36 | new byte[] | file://:0:0:0:0 | byte[] | file://:0:0:0:0 | byte |
| arrayCreations.kt:14:14:14:52 | new boolean[] | file://:0:0:0:0 | boolean[] | file://:0:0:0:0 | boolean |
| arrayCreations.kt:16:14:16:68 | new int[][] | file://:0:0:0:0 | int[][] | file://:0:0:0:0 | int[] |
| arrayCreations.kt:16:22:16:43 | new int[] | file://:0:0:0:0 | int[] | file://:0:0:0:0 | int |
| arrayCreations.kt:16:46:16:67 | new int[] | file://:0:0:0:0 | int[] | file://:0:0:0:0 | int |
| arrayCreations.kt:17:14:17:65 | new Object[] | file://:0:0:0:0 | Object[] | file://:0:0:0:0 | Object |
| arrayCreations.kt:17:22:17:43 | new int[] | file://:0:0:0:0 | int[] | file://:0:0:0:0 | int |
| arrayCreations.kt:17:46:17:64 | new Integer[] | file://:0:0:0:0 | Integer[] | file://:0:0:0:0 | Integer |
| arrayCreations.kt:18:14:18:62 | new Integer[][] | file://:0:0:0:0 | Integer[][] | file://:0:0:0:0 | Integer[] |
| arrayCreations.kt:18:22:18:40 | new Integer[] | file://:0:0:0:0 | Integer[] | file://:0:0:0:0 | Integer |
| arrayCreations.kt:18:43:18:61 | new Integer[] | file://:0:0:0:0 | Integer[] | file://:0:0:0:0 | Integer |
arrayCreationDimensions
| arrayCreations.kt:5:14:5:33 | new Integer[] | arrayCreations.kt:5:32:5:32 | 1 | 0 |
| arrayCreations.kt:6:14:6:32 | new Integer[] | arrayCreations.kt:6:14:6:32 | 4 | 0 |
| arrayCreations.kt:7:14:7:46 | new double[] | arrayCreations.kt:7:14:7:46 | 4 | 0 |
| arrayCreations.kt:8:14:8:49 | new float[] | arrayCreations.kt:8:14:8:49 | 4 | 0 |
| arrayCreations.kt:9:14:9:36 | new long[] | arrayCreations.kt:9:14:9:36 | 4 | 0 |
| arrayCreations.kt:10:14:10:35 | new int[] | arrayCreations.kt:10:14:10:35 | 4 | 0 |
| arrayCreations.kt:11:14:11:34 | new char[] | arrayCreations.kt:11:14:11:34 | 2 | 0 |
| arrayCreations.kt:12:14:12:37 | new short[] | arrayCreations.kt:12:14:12:37 | 4 | 0 |
| arrayCreations.kt:13:14:13:36 | new byte[] | arrayCreations.kt:13:14:13:36 | 4 | 0 |
| arrayCreations.kt:14:14:14:52 | new boolean[] | arrayCreations.kt:14:14:14:52 | 4 | 0 |
| arrayCreations.kt:16:14:16:68 | new int[][] | arrayCreations.kt:16:14:16:68 | 2 | 0 |
| arrayCreations.kt:16:22:16:43 | new int[] | arrayCreations.kt:16:22:16:43 | 4 | 0 |
| arrayCreations.kt:16:46:16:67 | new int[] | arrayCreations.kt:16:46:16:67 | 4 | 0 |
| arrayCreations.kt:17:14:17:65 | new Object[] | arrayCreations.kt:17:14:17:65 | 2 | 0 |
| arrayCreations.kt:17:22:17:43 | new int[] | arrayCreations.kt:17:22:17:43 | 4 | 0 |
| arrayCreations.kt:17:46:17:64 | new Integer[] | arrayCreations.kt:17:46:17:64 | 4 | 0 |
| arrayCreations.kt:18:14:18:62 | new Integer[][] | arrayCreations.kt:18:14:18:62 | 2 | 0 |
| arrayCreations.kt:18:22:18:40 | new Integer[] | arrayCreations.kt:18:22:18:40 | 4 | 0 |
| arrayCreations.kt:18:43:18:61 | new Integer[] | arrayCreations.kt:18:43:18:61 | 4 | 0 |
arrayCreationInit
| arrayCreations.kt:6:14:6:32 | new Integer[] | arrayCreations.kt:6:14:6:32 | {...} | arrayCreations.kt:6:22:6:22 | 1 | 0 |
| arrayCreations.kt:6:14:6:32 | new Integer[] | arrayCreations.kt:6:14:6:32 | {...} | arrayCreations.kt:6:25:6:25 | 2 | 1 |
| arrayCreations.kt:6:14:6:32 | new Integer[] | arrayCreations.kt:6:14:6:32 | {...} | arrayCreations.kt:6:28:6:28 | 3 | 2 |
| arrayCreations.kt:6:14:6:32 | new Integer[] | arrayCreations.kt:6:14:6:32 | {...} | arrayCreations.kt:6:31:6:31 | 4 | 3 |
| arrayCreations.kt:7:14:7:46 | new double[] | arrayCreations.kt:7:14:7:46 | {...} | arrayCreations.kt:7:28:7:30 | 1.0 | 0 |
| arrayCreations.kt:7:14:7:46 | new double[] | arrayCreations.kt:7:14:7:46 | {...} | arrayCreations.kt:7:33:7:35 | 2.0 | 1 |
| arrayCreations.kt:7:14:7:46 | new double[] | arrayCreations.kt:7:14:7:46 | {...} | arrayCreations.kt:7:38:7:40 | 3.0 | 2 |
| arrayCreations.kt:7:14:7:46 | new double[] | arrayCreations.kt:7:14:7:46 | {...} | arrayCreations.kt:7:43:7:45 | 4.0 | 3 |
| arrayCreations.kt:8:14:8:49 | new float[] | arrayCreations.kt:8:14:8:49 | {...} | arrayCreations.kt:8:27:8:30 | 1.0 | 0 |
| arrayCreations.kt:8:14:8:49 | new float[] | arrayCreations.kt:8:14:8:49 | {...} | arrayCreations.kt:8:33:8:36 | 2.0 | 1 |
| arrayCreations.kt:8:14:8:49 | new float[] | arrayCreations.kt:8:14:8:49 | {...} | arrayCreations.kt:8:39:8:42 | 3.0 | 2 |
| arrayCreations.kt:8:14:8:49 | new float[] | arrayCreations.kt:8:14:8:49 | {...} | arrayCreations.kt:8:45:8:48 | 4.0 | 3 |
| arrayCreations.kt:9:14:9:36 | new long[] | arrayCreations.kt:9:14:9:36 | {...} | arrayCreations.kt:9:26:9:26 | 1 | 0 |
| arrayCreations.kt:9:14:9:36 | new long[] | arrayCreations.kt:9:14:9:36 | {...} | arrayCreations.kt:9:29:9:29 | 2 | 1 |
| arrayCreations.kt:9:14:9:36 | new long[] | arrayCreations.kt:9:14:9:36 | {...} | arrayCreations.kt:9:32:9:32 | 3 | 2 |
| arrayCreations.kt:9:14:9:36 | new long[] | arrayCreations.kt:9:14:9:36 | {...} | arrayCreations.kt:9:35:9:35 | 4 | 3 |
| arrayCreations.kt:10:14:10:35 | new int[] | arrayCreations.kt:10:14:10:35 | {...} | arrayCreations.kt:10:25:10:25 | 1 | 0 |
| arrayCreations.kt:10:14:10:35 | new int[] | arrayCreations.kt:10:14:10:35 | {...} | arrayCreations.kt:10:28:10:28 | 2 | 1 |
| arrayCreations.kt:10:14:10:35 | new int[] | arrayCreations.kt:10:14:10:35 | {...} | arrayCreations.kt:10:31:10:31 | 3 | 2 |
| arrayCreations.kt:10:14:10:35 | new int[] | arrayCreations.kt:10:14:10:35 | {...} | arrayCreations.kt:10:34:10:34 | 4 | 3 |
| arrayCreations.kt:11:14:11:34 | new char[] | arrayCreations.kt:11:14:11:34 | {...} | arrayCreations.kt:11:26:11:28 | a | 0 |
| arrayCreations.kt:11:14:11:34 | new char[] | arrayCreations.kt:11:14:11:34 | {...} | arrayCreations.kt:11:31:11:33 | b | 1 |
| arrayCreations.kt:12:14:12:37 | new short[] | arrayCreations.kt:12:14:12:37 | {...} | arrayCreations.kt:12:27:12:27 | 1 | 0 |
| arrayCreations.kt:12:14:12:37 | new short[] | arrayCreations.kt:12:14:12:37 | {...} | arrayCreations.kt:12:30:12:30 | 2 | 1 |
| arrayCreations.kt:12:14:12:37 | new short[] | arrayCreations.kt:12:14:12:37 | {...} | arrayCreations.kt:12:33:12:33 | 3 | 2 |
| arrayCreations.kt:12:14:12:37 | new short[] | arrayCreations.kt:12:14:12:37 | {...} | arrayCreations.kt:12:36:12:36 | 4 | 3 |
| arrayCreations.kt:13:14:13:36 | new byte[] | arrayCreations.kt:13:14:13:36 | {...} | arrayCreations.kt:13:26:13:26 | 1 | 0 |
| arrayCreations.kt:13:14:13:36 | new byte[] | arrayCreations.kt:13:14:13:36 | {...} | arrayCreations.kt:13:29:13:29 | 2 | 1 |
| arrayCreations.kt:13:14:13:36 | new byte[] | arrayCreations.kt:13:14:13:36 | {...} | arrayCreations.kt:13:32:13:32 | 3 | 2 |
| arrayCreations.kt:13:14:13:36 | new byte[] | arrayCreations.kt:13:14:13:36 | {...} | arrayCreations.kt:13:35:13:35 | 4 | 3 |
| arrayCreations.kt:14:14:14:52 | new boolean[] | arrayCreations.kt:14:14:14:52 | {...} | arrayCreations.kt:14:29:14:32 | true | 0 |
| arrayCreations.kt:14:14:14:52 | new boolean[] | arrayCreations.kt:14:14:14:52 | {...} | arrayCreations.kt:14:35:14:39 | false | 1 |
| arrayCreations.kt:14:14:14:52 | new boolean[] | arrayCreations.kt:14:14:14:52 | {...} | arrayCreations.kt:14:42:14:45 | true | 2 |
| arrayCreations.kt:14:14:14:52 | new boolean[] | arrayCreations.kt:14:14:14:52 | {...} | arrayCreations.kt:14:48:14:51 | true | 3 |
| arrayCreations.kt:16:14:16:68 | new int[][] | arrayCreations.kt:16:14:16:68 | {...} | arrayCreations.kt:16:22:16:43 | new int[] | 0 |
| arrayCreations.kt:16:14:16:68 | new int[][] | arrayCreations.kt:16:14:16:68 | {...} | arrayCreations.kt:16:46:16:67 | new int[] | 1 |
| arrayCreations.kt:16:22:16:43 | new int[] | arrayCreations.kt:16:22:16:43 | {...} | arrayCreations.kt:16:33:16:33 | 1 | 0 |
| arrayCreations.kt:16:22:16:43 | new int[] | arrayCreations.kt:16:22:16:43 | {...} | arrayCreations.kt:16:36:16:36 | 2 | 1 |
| arrayCreations.kt:16:22:16:43 | new int[] | arrayCreations.kt:16:22:16:43 | {...} | arrayCreations.kt:16:39:16:39 | 3 | 2 |
| arrayCreations.kt:16:22:16:43 | new int[] | arrayCreations.kt:16:22:16:43 | {...} | arrayCreations.kt:16:42:16:42 | 4 | 3 |
| arrayCreations.kt:16:46:16:67 | new int[] | arrayCreations.kt:16:46:16:67 | {...} | arrayCreations.kt:16:57:16:57 | 1 | 0 |
| arrayCreations.kt:16:46:16:67 | new int[] | arrayCreations.kt:16:46:16:67 | {...} | arrayCreations.kt:16:60:16:60 | 2 | 1 |
| arrayCreations.kt:16:46:16:67 | new int[] | arrayCreations.kt:16:46:16:67 | {...} | arrayCreations.kt:16:63:16:63 | 3 | 2 |
| arrayCreations.kt:16:46:16:67 | new int[] | arrayCreations.kt:16:46:16:67 | {...} | arrayCreations.kt:16:66:16:66 | 4 | 3 |
| arrayCreations.kt:17:14:17:65 | new Object[] | arrayCreations.kt:17:14:17:65 | {...} | arrayCreations.kt:17:22:17:43 | new int[] | 0 |
| arrayCreations.kt:17:14:17:65 | new Object[] | arrayCreations.kt:17:14:17:65 | {...} | arrayCreations.kt:17:46:17:64 | new Integer[] | 1 |
| arrayCreations.kt:17:22:17:43 | new int[] | arrayCreations.kt:17:22:17:43 | {...} | arrayCreations.kt:17:33:17:33 | 1 | 0 |
| arrayCreations.kt:17:22:17:43 | new int[] | arrayCreations.kt:17:22:17:43 | {...} | arrayCreations.kt:17:36:17:36 | 2 | 1 |
| arrayCreations.kt:17:22:17:43 | new int[] | arrayCreations.kt:17:22:17:43 | {...} | arrayCreations.kt:17:39:17:39 | 3 | 2 |
| arrayCreations.kt:17:22:17:43 | new int[] | arrayCreations.kt:17:22:17:43 | {...} | arrayCreations.kt:17:42:17:42 | 4 | 3 |
| arrayCreations.kt:17:46:17:64 | new Integer[] | arrayCreations.kt:17:46:17:64 | {...} | arrayCreations.kt:17:54:17:54 | 1 | 0 |
| arrayCreations.kt:17:46:17:64 | new Integer[] | arrayCreations.kt:17:46:17:64 | {...} | arrayCreations.kt:17:57:17:57 | 2 | 1 |
| arrayCreations.kt:17:46:17:64 | new Integer[] | arrayCreations.kt:17:46:17:64 | {...} | arrayCreations.kt:17:60:17:60 | 3 | 2 |
| arrayCreations.kt:17:46:17:64 | new Integer[] | arrayCreations.kt:17:46:17:64 | {...} | arrayCreations.kt:17:63:17:63 | 4 | 3 |
| arrayCreations.kt:18:14:18:62 | new Integer[][] | arrayCreations.kt:18:14:18:62 | {...} | arrayCreations.kt:18:22:18:40 | new Integer[] | 0 |
| arrayCreations.kt:18:14:18:62 | new Integer[][] | arrayCreations.kt:18:14:18:62 | {...} | arrayCreations.kt:18:43:18:61 | new Integer[] | 1 |
| arrayCreations.kt:18:22:18:40 | new Integer[] | arrayCreations.kt:18:22:18:40 | {...} | arrayCreations.kt:18:30:18:30 | 1 | 0 |
| arrayCreations.kt:18:22:18:40 | new Integer[] | arrayCreations.kt:18:22:18:40 | {...} | arrayCreations.kt:18:33:18:33 | 2 | 1 |
| arrayCreations.kt:18:22:18:40 | new Integer[] | arrayCreations.kt:18:22:18:40 | {...} | arrayCreations.kt:18:36:18:36 | 3 | 2 |
| arrayCreations.kt:18:22:18:40 | new Integer[] | arrayCreations.kt:18:22:18:40 | {...} | arrayCreations.kt:18:39:18:39 | 4 | 3 |
| arrayCreations.kt:18:43:18:61 | new Integer[] | arrayCreations.kt:18:43:18:61 | {...} | arrayCreations.kt:18:51:18:51 | 1 | 0 |
| arrayCreations.kt:18:43:18:61 | new Integer[] | arrayCreations.kt:18:43:18:61 | {...} | arrayCreations.kt:18:54:18:54 | 2 | 1 |
| arrayCreations.kt:18:43:18:61 | new Integer[] | arrayCreations.kt:18:43:18:61 | {...} | arrayCreations.kt:18:57:18:57 | 3 | 2 |
| arrayCreations.kt:18:43:18:61 | new Integer[] | arrayCreations.kt:18:43:18:61 | {...} | arrayCreations.kt:18:60:18:60 | 4 | 3 |
